pkg/resource/edit/operations_test.go

```diff
@@ -91,6 +91,67 @@ func TestDeletion(t *testing.T) {
 	assert.Equal(t, []*resource.State{pA, a, c}, snap.Resources)
 }
 
+func TestDeletingDuplicateDependencies(t *testing.T) {
+	t.Parallel()
+
+	pA := NewProviderResource("a", "p1", "0")
+	a := NewResource("a", pA)
+
+	// Create duplicate resources.
+	b1 := NewResource("b", pA)
+	b2 := NewResource("b", pA)
+	b3 := NewResource("b", pA)
+
+	// ensure b1, b2, and b3 must have the same URN.
+	bURN := b1.URN
+	assert.Equal(t, bURN, b1.URN)
+	assert.Equal(t, bURN, b2.URN)
+	assert.Equal(t, bURN, b3.URN)
+
+	// c exists to check behavior on b's dependents.
+	c := NewResource("c", pA, bURN)
+
+	t.Run("do-target-dependents", func(t *testing.T) {
+		t.Parallel()
+		snap := NewSnapshot([]*resource.State{
+			pA, a, b1, b2, b3, c,
+		})
+
+		err := DeleteResource(snap, b1, nil, true /* targetDependents */)
+		require.NoError(t, err)
+
+		// assert.Equals does a deep equals with the expected list.
+		assert.Equal(t, []*resource.State{
+			pA, a, b2, b3, c,
+		}, snap.Resources)
+
+		// Ensure that a pointer to b1 is not in the list.
+		for _, s := range snap.Resources {
+			assert.False(t, s == b1)
+		}
+	})
+
+	t.Run("do-not-target-dependents", func(t *testing.T) {
+		t.Parallel()
+		snap := NewSnapshot([]*resource.State{
+			pA, a, b1, b2, b3, c,
+		})
+
+		err := DeleteResource(snap, b1, nil, false /* targetDependents */)
+		require.NoError(t, err)
+
+		// assert.Equals does a deep equals with the expected list.
+		assert.Equal(t, []*resource.State{
+			pA, a, b2, b3, c,
```

Comment:
I'm not sure how these two test cases are different in terms of behavior tested.
The `true`/`false` on targetDependents doesn't change the rest of the test.
Would it be more useful to delete something with dependents in do-target-dependents?

Optionally, I would also consider a test with targetDependents false that tries to delete pA,
and `assert.ErrorAs` the result:

    err := DeleteResource(snap, pA, nil, false /* targetDependents */)
    assert.ErrorAs(t, err, &ResourceHasDependenciesError{})

Reply:
There are some existing tests on this code so I didn't consider adding too many more.

Given the motivation for this PR was #12077, I want to make sure that the command is successful when targetDependents is false.

`do-target-dependents` does potentially target c because c depends on the URN that b1, b2, and b3 all share.

thx for the very thought provoking question. I will add a comment to clarify intent and context around the addition of these tests.